presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -252,7 +260,53 @@ public static boolean canCoerce(Type actualType, Type expectedType)
     public static boolean canCoerce(TypeSignature actualType, TypeSignature expectedType)
     {
         Optional<TypeSignature> commonSuperTypeSignature = getCommonSuperTypeSignature(actualType, expectedType);
-        return commonSuperTypeSignature.isPresent() && commonSuperTypeSignature.get().equals(expectedType);
+        return commonSuperTypeSignature.isPresent() && typesMatch(commonSuperTypeSignature.get(), expectedType);
+    }
+
+    private static boolean typesMatch(TypeSignature actualType, TypeSignature expectedType)
```

Comment:
In order to make literal parameters to calculate on a fly we need to refactor all `matchAndBind` code. I think it is better to do as you said, as a separate PR.